shell/core/src/main/java/org/apache/karaf/shell/impl/console/ConsoleSessionImpl.java

```diff
@@ -451,7 +449,12 @@ private CharSequence readCommand(AtomicBoolean reading) throws UserInterruptExce
                 command = reader.getBuffer().toString();
             }
         } catch (EndOfFileException e) {
-            command = null;
+            boolean disableEofExit = (boolean) session.get(Session.DISABLE_EOF_EXIT);
+            if (disableEofExit) {
+                command = "";
```

Comment:
I tested it manually and it works fine but I couldn't figure out a way to write a test for this (e.g. how to send CTRL-D to Karaf in a pax exam test, or some other way)

Reply:
That's actually pretty hard in pax-exam as it controls the lifecycle. Manual test is fine and enough IMHO.